drivers/android/process.rs

```diff
@@ -795,8 +795,9 @@ impl IoctlHandler for Process {
         data: UserSlicePtr,
     ) -> Result<i32> {
         let thread = this.get_thread(Task::current().pid())?;
+        let blocking = (file.flags().flags() & FileFlags::NON_BLOCK) == 0;
```

Comment:
I don't think the ergonomics of`.flags().flags()` is great.

I struggled a bit with this: on one hand, it is better to return a type so that the type system is informing (and enforcing) what the values are; on another hand, having to force usage patterns like the one above isn't great.

I opted for using the equivalent of `FileFlags` elsewhere as a mere container of constants, a namespace if you will. I could be convinced otherwise, but for now I'm thinking that that's what it should be, just a namespace for constants.

Reply:
Any reason against bitflags-style wrapper apart from difficulty of usage in const eval?

Reply:
This seems like a good middleground between importing the bitflags crate and hand-implementing bitflags style functionality for each set of flags we expose.

I would personally prefer importing bitflags crate but I can understand why maintainers would not want to.

Reply:
The code part of `bitflags` is just like a few hundreds lines. I think we could have a similar macro vendored in.

Reply:
If there is consensus for that I can help get that in.

Reply:
I haven't used `bitflags` in the past. My impression looking at it now is that the syntax seems reasonable.

One concern I have is about the `!-operator`: IIUC, it's a set complement, so in the [example](https://docs.rs/bitflags/latest/bitflags/), !6 is 1 instead of 0xfffffff9 (which is what I would expect from a `u32`). In addition to not seeming conventional (to me at least), this may be problematic if/when we get out of sync with the C side (e.g., a new flag is introduced there but not in Rust).

I wouldn't be in favor of having unconventional operators. In my experience, unexpected behaviour is a recipe for bugs.

Reply:
Excluding tests and docs the entire bitflags crate is ~600 lines of code. Vendoring it and removing the `Not` impl (https://docs.rs/bitflags/latest/src/bitflags/lib.rs.html#798) and maybe the `complement` function (which does the same) shouldn't be too hard.

Reply:
It also seems that `rust-bindgen` has an option for generating a "light" version of `bitflags` using `--bitfield-enum`: https://github.com/rust-lang/rust-bindgen/pull/226

Reply:
`bitflags` has an invariant that the flags it generates are exhaustive; therefore the `!` operator it provides is logical and consistent. `0xfffffff9` would not be representable.

If we want non-exhaustive bitflags then we need to make a few changes; the invariant could be removed along with from_bits_truncate/from_bits_unchecked/all/etc. We could add a helper type:
```rust
struct Mask<T>(T);

impl Not for Flags {
	type Output = Mask<Self>;
	fn not(self) -> Self::Output {
		Mask(Self(!self.0))
	}
}

impl BitAnd<Mask<Self>> for Flags {
	type Output = Flags;
	fn bit_and(self, other: Mask<Self>) -> Self::Output {
		Self(self.0 & other.0.0)
	}
}
```